examples/subplots_axes_and_figures/subplots_demo.py

```diff
@@ -1,103 +1,192 @@
+ax2.plot(x, y**2, 'tab:orange')
+ax3.plot(x, -y, 'tab:green')
+ax4.plot(x, -y**2, 'tab:red')
+
+for ax in axs.flat:
+    ax.label_outer()
 
 ###############################################################################
-# Four axes, returned as a 2-d array
-
-f, axarr = plt.subplots(2, 2)
-axarr[0, 0].plot(x, y)
-axarr[0, 0].set_title('Axis [0,0]')
-axarr[0, 1].scatter(x, y)
-axarr[0, 1].set_title('Axis [0,1]')
-axarr[1, 0].plot(x, y ** 2)
-axarr[1, 0].set_title('Axis [1,0]')
-axarr[1, 1].scatter(x, y ** 2)
-axarr[1, 1].set_title('Axis [1,1]')
-for ax in axarr.flat:
-    ax.set(xlabel='x-label', ylabel='y-label')
-# Hide x labels and tick labels for top plots and y ticks for right plots.
-for ax in axarr.flat:
+# Sharing axes
+# """"""""""""
+#
+# By default, each Axes is scaled individually. Thus, if the ranges are
+# different the tick values of the subplots do not align.
+
+fig, (ax1, ax2) = plt.subplots(2)
+fig.suptitle('Axes values are scaled individually by default')
+ax1.plot(x, y)
+ax2.plot(x + 1, -y)
+
+###############################################################################
+#
+# You can use *sharex* or *sharey* to align the horizontal or vertical axis.
+
+fig, (ax1, ax2) = plt.subplots(2, sharex=True)
+fig.suptitle('Aligning x-axis using sharex')
+ax1.plot(x, y)
+ax2.plot(x + 1, -y)
+
+###############################################################################
+# Setting *sharex* or *sharey* to ``True`` enables global sharing across the
+# whole grid, i.e. also the y-axes of vertically stacked subplots have the
+# same scale when using ``sharey=True``.
+
+fig, axs = plt.subplots(3, sharex=True, sharey=True)
+fig.suptitle('Sharing both axes')
+axs[0].plot(x, y ** 2)
+axs[1].plot(x, 0.3 * y, 'o')
+axs[2].plot(x, y, '+')
+
+###############################################################################
+# For subplots that are sharing axes one set of tick labels is enough. Tick
+# labels of inner Axes are automatically removed by *sharex* and *sharey*.
+# Still there remains an unused empty space between the subplots.
+#
+# The parameter *gridspec_kw* of `.pyplot.subplots` controls the grid
+# properties (see also `.GridSpec`). For example, we can reduce the height
+# between vertical subplots using ``gridspec_kw={'hspace': 0}``.
+#
+# `.label_outer` is a handy method to remove labels and ticks from subplots
+# that are not at the edge of the grid.
+
+fig, axs = plt.subplots(3, sharex=True, sharey=True, gridspec_kw={'hspace': 0})
+fig.suptitle('Sharing both axes')
+axs[0].plot(x, y ** 2)
+axs[1].plot(x, 0.3 * y, 'o')
+axs[2].plot(x, y, '+')
+
+# Hide x labels and tick labels for all but bottom plot.
+for ax in axs:
+    ax.label_outer()
+
+###############################################################################
+# Apart from ``True`` and ``False``, both *sharex* and *sharey* accept the
+# values 'row' and 'col' to share the values only per row or column.
+
+fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2,
```

Comment:
In general I don't like very short lines all squeezed to the right, I'd suggest e.g.
```
fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(
    2, 3, sharex='col', sharey='row',
    gridspec_kw={...})
```